### PR TITLE
Laboratory work #3, Menyalshchikov Timur - 21FPL1

### PR DESCRIPTION
**Pull request by Timur20033**

### CODE REVIEW

File: lab_3_keywords_textrank/main.py [hunk 1]
```diff
@@ -81,7 +90,7 @@ def preprocess_text(self, text: str) -> tuple[str, ...]:
             tuple[str, ...]
                 clean lowercase tokens with no stop-words
         """
-        pass
+        return self._remove_stop_words(self._clean_and_tokenize(text))
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:93]: please, save the result of `self._clean_and_tokenize(text)` to a variable

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_keywords_textrank/main.py (hunk 1 above)
Comment (Timur20033) [reply lab_3_keywords_textrank/main.py]: done

File: lab_3_keywords_textrank/main.py [hunk 2]
```diff
@@ -66,7 +71,11 @@ def _remove_stop_words(self, tokens: tuple[str, ...]) -> tuple[str, ...]:
             tuple[str, ...]
                 tokens without stop-words
         """
-        pass
+        words = []
+        for element in tokens:
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:75]: Can you write it in a single line?

File: lab_3_keywords_textrank/main.py [hunk 3]
```diff
@@ -120,7 +131,13 @@ def _learn_indices(self, tokens: tuple[str, ...]) -> None:
             tokens : tuple[str, ...]
                 sequence of string tokens
         """
-        pass
+        unique = []
+        for element in tokens:
+            if element not in unique:
+                unique.append(element)
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:137]: Have you heard about `set`?

File: lab_3_keywords_textrank/main.py [hunk 4]
```diff
@@ -120,7 +131,13 @@ def _learn_indices(self, tokens: tuple[str, ...]) -> None:
...
+        for idx, element in enumerate(unique, start=1000):
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:138]: perfect

File: lab_3_keywords_textrank/main.py [hunk 5]
```diff
@@ -136,7 +153,10 @@ def encode(self, tokens: tuple[str, ...]) -> Optional[tuple[int, ...]]:
                 sequence of integer tokens
         In case of empty tokens input data, None is returned
         """
-        pass
+        if not tokens:
+            return None
+        self._learn_indices(tokens)
+        return tuple(self._word2id.values())
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:159]: Are you sure that it you need just values?

File: lab_3_keywords_textrank/main.py [hunk 6]
```diff
@@ -152,7 +172,12 @@ def decode(self, encoded_tokens: tuple[int, ...]) -> Optional[tuple[str, ...]]:
                 sequence of string tokens
         In case of out-of-dictionary input data, None is returned
         """
-        pass
+        if any(number not in self._id2word.keys() for number in encoded_tokens):
+            return None
+        words = []
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:177]: again - one line

File: lab_3_keywords_textrank/main.py [hunk 7]
```diff
@@ -173,7 +198,21 @@ def extract_pairs(tokens: tuple[int, ...], window_length: int) -> Optional[tuple
     In case of corrupt input data, None is returned:
     tokens must not be empty, window lengths must be integer, window lengths cannot be less than 2.
     """
-    pass
+    if not tokens or not isinstance(window_length, int) or window_length < 2:
+        return None
+    pairs = []
+    for num in range(0, len(tokens) - window_length + 1):
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:204]: looks like it is a very suboptimal

File: lab_3_keywords_textrank/main.py [hunk 8]
```diff
@@ -233,7 +274,21 @@ def add_edge(self, vertex1: int, vertex2: int) -> int:
                 0 if edge was added successfully, otherwise -1
         In case of vertex1 being equal to vertex2, -1 is returned as loops are prohibited
         """
-        pass
+        if vertex1 == vertex2:
+            return -1
+        if self._matrix == [[]]:
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py]: I believe, you can do it inside `init`

File: lab_3_keywords_textrank/main.py [hunk 9]
```diff
@@ -233,7 +274,21 @@ def add_edge(self, vertex1: int, vertex2: int) -> int:
...
+            self._matrix[0].append(0)
+        for vertex in [vertex1, vertex2]:
+            if vertex not in self._matrix[0]:
+                self._matrix[0].append(vertex)
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py]: What exactly does `self._matrix[0]` store?

File: lab_3_keywords_textrank/main.py [hunk 10]
```diff
@@ -233,7 +275,18 @@ def add_edge(self, vertex1: int, vertex2: int) -> int:
                 0 if edge was added successfully, otherwise -1
         In case of vertex1 being equal to vertex2, -1 is returned as loops are prohibited
         """
-        pass
+        if vertex1 == vertex2:
+            return -1
+        for vertex in [vertex1, vertex2]:
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:280]: I hope, you can use: `for vertex in vertex1, vertex2:`

File: lab_3_keywords_textrank/main.py [hunk 11]
```diff
@@ -233,7 +275,18 @@ def add_edge(self, vertex1: int, vertex2: int) -> int:
...
+            if vertex not in self._vertices:
+                self._vertices.append(vertex)
+        self._matrix.extend([[] for _ in range(len(self._vertices) - len(self._matrix))])
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:283]: I suggest moving this to a help method

File: lab_3_keywords_textrank/main.py [hunk 12]
```diff
@@ -251,7 +304,11 @@ def is_incidental(self, vertex1: int, vertex2: int) -> int:
                 1 if vertices are incidental, otherwise 0
         If either of vertices is not present in the graph, -1 is returned
         """
-        pass
+        if vertex1 in self._vertices and vertex2 in self._vertices:
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:307]: revert and return immediately

File: lab_3_keywords_textrank/main.py [hunk 13]
```diff
@@ -278,7 +335,11 @@ def calculate_inout_score(self, vertex: int) -> int:
                 number of incidental vertices
         If vertex is not present in the graph, -1 is returned
         """
-        pass
+        if vertex in self._vertices:
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:338]: again

File: lab_3_keywords_textrank/main.py [hunk 14]
```diff
@@ -302,14 +365,24 @@ def fill_positions(self, tokens: tuple[int, ...]) -> None:
             tokens : tuple[int, ...]
                 sequence of tokens
         """
-        pass
+        for idx, element in enumerate(tokens):
+            if element not in self._positions.keys():
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:369]: you can remove `.keys()`

File: lab_3_keywords_textrank/main.py [hunk 15]
```diff
@@ -302,14 +365,24 @@ def fill_positions(self, tokens: tuple[int, ...]) -> None:
...
         """
-        pass
+        dct = {}
+        for vertex in self._positions:
+            summary = 0.0
+            for position in self._positions[vertex]:
+                summary += 1/position
+            dct[vertex] = summary
+        for element in dct:
+            self._position_weights[element] = dct[element]/sum(dct.values())
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:385]: calculate  ` dct[element]/sum(dct.values())` once

File: lab_3_keywords_textrank/main.py [hunk 16]
```diff
@@ -302,14 +365,24 @@ def fill_positions(self, tokens: tuple[int, ...]) -> None:
...
+                self._positions[element] = []
+            self._positions[element].append(idx + 1)
 
     # Step 8.3
     def calculate_position_weights(self) -> None:
         """
         Computes position weights for all tokens in text
         """
-        pass
+        dct = {}
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:378]: very bad name for this variable

File: lab_3_keywords_textrank/main.py [hunk 17]
```diff
@@ -320,7 +393,8 @@ def get_position_weights(self) -> dict[int, float]:
             dict[int, float]
                 position weights for all vertices in the graph
         """
-        pass
+        weights = self._position_weights
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:396]: ? just return?

File: lab_3_keywords_textrank/main.py [hunk 18]
```diff
@@ -419,7 +508,10 @@ def calculate_inout_score(self, vertex: int) -> int:
                 number of incidental vertices
         If vertex is not present in the graph, -1 is returned
         """
-        pass
+        if vertex not in self._edges.keys():
+            return -1
+        result = len(self._edges[vertex])
+        return result
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:514]: why do you save it to a variable?

File: lab_3_keywords_textrank/main.py [hunk 19]
```diff
@@ -461,7 +565,8 @@ def get_position_weights(self) -> dict[int, float]:
             dict[int, float]
                 position weights for all vertices in the graph
         """
-        pass
+        weights = self._position_weights
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:568]: again

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_keywords_textrank/main.py (hunk 2 above)
Comment (Timur20033) [reply lab_3_keywords_textrank/main.py:75]: done

File: lab_3_keywords_textrank/main.py (hunk 5 above)
Comment (Timur20033) [reply lab_3_keywords_textrank/main.py:159]: done

File: lab_3_keywords_textrank/main.py (hunk 6 above)
Comment (Timur20033) [reply lab_3_keywords_textrank/main.py:177]: done

File: lab_3_keywords_textrank/main.py (hunk 7 above)
Comment (Timur20033) [reply lab_3_keywords_textrank/main.py:204]: done

File: lab_3_keywords_textrank/main.py (hunk 9 above)
Comment (Timur20033) [reply lab_3_keywords_textrank/main.py]: done

File: lab_3_keywords_textrank/main.py (hunk 10 above)
Comment (Timur20033) [reply lab_3_keywords_textrank/main.py:280]: done

File: lab_3_keywords_textrank/main.py (hunk 11 above)
Comment (Timur20033) [reply lab_3_keywords_textrank/main.py]: Could you tell, what it means, please?

File: lab_3_keywords_textrank/main.py (hunk 12 above)
Comment (Timur20033) [reply lab_3_keywords_textrank/main.py]: done

File: lab_3_keywords_textrank/main.py (hunk 13 above)
Comment (Timur20033) [reply lab_3_keywords_textrank/main.py]: done

File: lab_3_keywords_textrank/main.py (hunk 14 above)
Comment (Timur20033) [reply lab_3_keywords_textrank/main.py:367]: done

File: lab_3_keywords_textrank/main.py (hunk 16 above)
Comment (Timur20033) [reply lab_3_keywords_textrank/main.py:376]: done

File: lab_3_keywords_textrank/main.py (hunk 17 above)
Comment (Timur20033) [reply lab_3_keywords_textrank/main.py:394]: yes...)

File: lab_3_keywords_textrank/main.py (hunk 18 above)
Comment (Timur20033) [reply lab_3_keywords_textrank/main.py:512]: done

File: lab_3_keywords_textrank/main.py (hunk 19 above)
Comment (Timur20033) [reply lab_3_keywords_textrank/main.py:566]: done

File: lab_3_keywords_textrank/main.py [hunk 20]
```diff
@@ -36,7 +36,8 @@ def __init__(self, stop_words: tuple[str, ...], punctuation: tuple[str, ...]) ->
             punctuation : tuple[str, ...]
                 punctuation symbols to remove during text cleaning
         """
-        pass
+        self.stop_words = stop_words
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:39]: hide this attributes using `_`

File: lab_3_keywords_textrank/main.py [hunk 21]
```diff
@@ -152,7 +168,9 @@ def decode(self, encoded_tokens: tuple[int, ...]) -> Optional[tuple[str, ...]]:
                 sequence of string tokens
         In case of out-of-dictionary input data, None is returned
         """
-        pass
+        if any(number not in self._id2word.keys() for number in encoded_tokens):
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:171]: remove .keys()

File: lab_3_keywords_textrank/main.py (hunk 3 above)
Comment (artyomtugaryov) [reply lab_3_keywords_textrank/main.py:133]: why does not decode work?

File: lab_3_keywords_textrank/main.py [hunk 22]
```diff
@@ -173,7 +191,21 @@ def extract_pairs(tokens: tuple[int, ...], window_length: int) -> Optional[tuple
...
+    if not tokens or not isinstance(window_length, int) or window_length < 2:
+        return None
+    pairs = []
+    for num in range(len(tokens) + 1 - window_length):
+        window = tokens[num:window_length + num]
+        for element in window:
+            small = window[window.index(element) + 1:]
+            for subel in small:
+                pairs.append([element, subel])
+    unique = []
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:203]: again - what about set?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_keywords_textrank/main.py (hunk 21 above)
Comment (Timur20033) [reply lab_3_keywords_textrank/main.py:171]: done

File: lab_3_keywords_textrank/main.py (hunk 20 above)
Comment (Timur20033) [reply lab_3_keywords_textrank/main.py:39]: done

File: lab_3_keywords_textrank/main.py (hunk 22 above)
Comment (Timur20033) [reply lab_3_keywords_textrank/main.py:203]: I tried using set, now all works somehow... the second case I also did

**Review by artyomtugaryov — APPROVED** [review]

**Pull request closed without merging**